core/isolate.rs

```diff
@@ -52,9 +54,33 @@ pub enum StartupData<'a> {
 
 type DispatchFn = Fn(&[u8], Option<PinnedBuf>) -> Op;
 
+pub type DynImportFuture = Box<dyn Future<Item = deno_mod, Error = ()> + Send>;
+type DynImportFn = Fn(&str, &str) -> DynImportFuture;
+
+/// Wraps DynImportFuture to include the deno_dyn_import_id, so that it doesn't
+/// need to be exposed.
+struct DynImport {
+  id: deno_dyn_import_id,
+  inner: DynImportFuture,
+}
+
+impl Future for DynImport {
+  type Item = (deno_dyn_import_id, deno_mod);
+  type Error = ();
+  fn poll(&mut self) -> Poll<Self::Item, ()> {
```

Comment:
A little unfortunate that this is necessary, since this is just a complicated way of writing `dyn_import_inner(...).map(|mod_id| (id, mod_id)).or_else(|_| Ok(id, 0))`.
I guess it's necessary to give the `T` in `FuturesUnordered<T>` a concrete type?

Reply:
It's just to hide the id from the user.